android/src/main/java/com/rt2zz/reactnativecontacts/ContactsProvider.java

```diff
@@ -0,0 +1,181 @@
+package com.rt2zz.reactnativecontacts;
+
+import android.content.ContentResolver;
+import android.database.Cursor;
+import android.net.Uri;
+import android.provider.ContactsContract;
+import android.text.TextUtils;
+
+import com.facebook.react.bridge.Arguments;
+import com.facebook.react.bridge.WritableArray;
+import com.facebook.react.bridge.WritableMap;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ContactsProvider {
+
+    private final Uri QUERY_URI = ContactsContract.Contacts.CONTENT_URI;
+    private final String CONTACT_ID = ContactsContract.Contacts._ID;
+    private final String DISPLAY_NAME = ContactsContract.Contacts.DISPLAY_NAME;
+    private final String PHOTO_URI = ContactsContract.CommonDataKinds.Contactables.PHOTO_URI;
+    private final Uri EMAIL_CONTENT_URI = ContactsContract.CommonDataKinds.Email.CONTENT_URI;
+    private final String EMAIL_CONTACT_ID = ContactsContract.CommonDataKinds.Email.CONTACT_ID;
+    private final String EMAIL_DATA = ContactsContract.CommonDataKinds.Email.DATA;
+    private final String HAS_PHONE_NUMBER = ContactsContract.Contacts.HAS_PHONE_NUMBER;
+    private final String PHONE_NUMBER = ContactsContract.CommonDataKinds.Phone.NUMBER;
+    private final Uri PHONE_CONTENT_URI = ContactsContract.CommonDataKinds.Phone.CONTENT_URI;
+    private final String PHONE_CONTACT_ID = ContactsContract.CommonDataKinds.Phone.CONTACT_ID;
+    private final String STARRED_CONTACT = ContactsContract.Contacts.STARRED;
+    private final ContentResolver contentResolver;
+
+    public ContactsProvider(ContentResolver contentResolver) {
+        this.contentResolver = contentResolver;
+    }
+
+    public WritableArray getContacts() {
+        List<Contact> contactList = new ArrayList<>();
+        String[] projection = new String[]{CONTACT_ID, DISPLAY_NAME, PHOTO_URI, HAS_PHONE_NUMBER, STARRED_CONTACT};
+        String selection = null;
+        Cursor cursor = contentResolver.query(QUERY_URI, projection, selection, null, null);
+
+        while (cursor.moveToNext()) {
+            contactList.add(createContact(cursor));
+        }
+
+        cursor.close();
+
+        WritableArray contacts = Arguments.createArray();
+        for (Contact contact : contactList) {
+            contacts.pushMap(contact.toMap());
+        }
+
+        return contacts;
+    }
+
+    private Contact createContact(Cursor cursor) {
+        int contactId = cursor.getInt(cursor.getColumnIndex(CONTACT_ID));
```

Comment:
I've done some research, and you can optimize the performance by getting the column indices only once, before the while loop. This way you're not looking it up for every contact.